### PR TITLE
[BEAM-1835] Make Pubsub ackBatch safer

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Checkpoints can be finalized at any point, including while the reader is
being closed or after it has been. When ackBatch is called and the
reader is closed, it should return. When close is called while a
different thread is using the client, it should wait for that thread to
complete before closing the reader.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi 

I've made sure that the client is not closed until all outstanding checkpoints are finalized, and added a test for this sequencing.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 1]
```diff
@@ -398,10 +407,16 @@ public void encode(PubsubCheckpoint<T> value, OutputStream outStream, Context co
...
     /**
-     * Client on which to talk to Pubsub. Null if closed.
+     * Client on which to talk to Pubsub. Contains a null value if closed.
      */
-    @Nullable
-    private PubsubClient pubsubClient;
+    private AtomicReference<PubsubClient> pubsubClient;
+
+    /**
+     * The number of currently active calls to {@link #ackBatch(List)}. Batches can be ACKed from
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: javadoc seems wrong -- it's a bool not a count.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 2]
```diff
@@ -626,17 +642,17 @@ public PubsubReader(PubsubOptions options, PubsubSource<T> outer, SubscriptionPa
...
-      return pubsubClient;
+      return pubsubClient.get();
     }
 
     /**
      * BLOCKING
      * ACK {@code ackIds} back to Pubsub.
-     * CAUTION: May be invoked from a separate checkpointing thread.
+     * CAUTION: May be invoked from a separate thread.
      * CAUTION: Retains {@code ackIds}.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: javadoc-ify this?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 3]
```diff
@@ -312,8 +315,14 @@ public void finalizeCheckpoint() throws IOException {
           reader.ackBatch(batchSafeToAckIds);
         }
       } finally {
-        checkState(reader.numInFlightCheckpoints.decrementAndGet() >= 0,
+        int remainingInFlight = reader.numInFlightCheckpoints.decrementAndGet();
+        checkState(remainingInFlight >= 0,
                    "Miscounted in-flight checkpoints");
+        if (!reader.active.get() && remainingInFlight == 0) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: optional suggest: move all the checks into `maybeCloseClient()` and just call it unconditionally upon finalize checkpoint or reader close.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: Fixed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: Done, I think.

**Comment by tgroh** [comment]
retest this please

**Pull request closed without merging**